### DISCUSSION TITLE
How to clear an element before ElemSetTxtStr (remove old text)

### DISCUSSION THREAD

**Discussion by PrattTechnologies** [ImpulseAdventure/GUIslice#404] [Q&A] [answered]
Hi

When I update an element with ElemSetTxtStr, if the old text was longer or 'taller' than the new text, then there are remnants of the old text even after the element is updated.
Is there a way to clear or set to zero an element so the old text is not visible, only the new text?

I have tried setting the element to blank spaces (i.e. in a 5 length element "     ") but this does not work as I want (to clear the element completely) unfortunately.

Thank you.

**Comment by Pconti31** [comment]
@PrattTechnologies If you add a zip file of a small sample project that reproduces your problem it makes coming up either what you are doing wrong or with a bug fix to GUIslice easy.  Just drag and drop the zip file anywhere in a new comment or answer.  If the discussions section doesn't work you may need to repost in the issues section to attach files.
Paul--

**Comment by ImpulseAdventure** [comment]
GUIslice should redraw the entire text widget if you call ElemSetTxtStr(). What you describe might suggest that your new text is larger than the size of the text widget specified in the Builder (assuming you are using the Builder). I would first try to increase the size of the text field in the Builder to match the largest string that you might set via ElemSetTxtStr().

As Paul suggests, attaching your project will help us determine what your code may be doing. From that, we should be able to infer the text strings that are being passed into ElemSetTxtStr().

thx

**Reply by PrattTechnologies** [reply]
Thanks @ImpulseAdventure @Pconti31 
Sorry for the large delay, I have attached a zip folder here with an example sketch and also some images that show what happens on the TFT. This happens using both MCUFriend on a 2.8" and UTFT on a 7" TFT.

Essentially the old text is not removed before it is overwritten so you can see the old text 'through' the new text. In the images I sent '4' and then '6'.
![tft1](https://user-images.githubusercontent.com/81346972/132967762-3f974933-8133-4794-b220-d73c8b8424a5.jpg)
![tft2](https://user-images.githubusercontent.com/81346972/132967763-52c735e5-b07c-490f-ad4f-ce526d0a2930.jpg)
![tft3](https://user-images.githubusercontent.com/81346972/132967764-ac22d697-63fc-43b0-8b2a-fc1d0c49a751.jpg)


Thanks for your help.
[guislice debug.zip](https://github.com/ImpulseAdventure/GUIslice/files/7148840/guislice.debug.zip)

**Comment by Pconti31** [comment, marked as answer]
@PrattTechnologies @ImpulseAdventure Well, I couldn't reproduce your problem but I found enough bugs in your program to point to possible issues.

1- You must call gslc_Update every cycle through loop() or your UI will become unresponsive.  Also, you need a delay in loop() to show all messages received or you will only see the last full message.
```
void loop() {
  // put your main code here, to run repeatedly:
  gslc_Update(&m_gui);
  delay(1000); 
  serialRead();
}
```
2- Inside serailRead you fail to terminate your message buffer during initalization so it will have garbage inside it and if it gets displayed you will see the results you posted.  

3- Your read loop as written can over-write your message buffer because you have no guard in place (i<4).

4- Now while you do have a snprintf of 4 you don't actually need to terminate your message buffer on each cycle so it not really a bug but its always good practice to do so anyway.  That's up to you.  In the sample code below I do so just to display in the serial monitor what is happening on each read.  Now one real bug here is your snprint is set to 4 and it should be 5.

So in summary I would code your seriallRead() as:
```
void serialRead()
{
  char serialMsg[5];
  strcpy(serialMsg, "");  // must initialize
  int i = 0;
  while (Serial.available() > 0 && i<4) {  // add guard
    serialMsg[i] = Serial.read();
    i++;
    serialMsg[i] = '\0'; // add trailing nul - strictly speaking only needed for following print
    Serial.println(serialMsg);
  }

  delay(10); // not sure why you addded this....

  if (strcmp(serialMsg, "")) {
  char acTxt[5];
  
  snprintf(acTxt, 5, serialMsg);
  gslc_ElemSetTxtStr(&m_gui, m_pElemInput, acTxt);
//  gslc_Update(&m_gui);
//  strcpy(serialMsg, "");
  }
}
```
Now, If these changes still leave you with a problem its possible its your font size.  Increase the height and width of your m_pElemInput.  From 91,26 to larger values say 110, 40

If all of this fails then I'm sorry but you will need help from Calvin.  This is as far as I can go.
Paul--

**Reply by PrattTechnologies** [reply]
Hi @Pconti31 
Thank you for your help!
I fixed the syntax but the thing that worked was changing the size of the element.

**Comment by Pconti31** [comment]
@PrattTechnologies One last point I forgot to mention.  You don't need to do a snprint to acTxt.  Simply:
```
//  snprintf(acTxt, 5, serialMsg);
//  gslc_ElemSetTxtStr(&m_gui, m_pElemInput, acTxt);
  gslc_ElemSetTxtStr(&m_gui, m_pElemInput, serialMsg);
```
Since inside your debug_GSLC.h the gslc_ElemCreateTxt() call set the max size of its buffer and will not allow any overwrite.
Paul--
